discovery: swap mdns lib for grandcat/zeroconf

rargulati opened the pull request:
Specifically addressing the request here: https://github.com/libp2p/go-libp2p/issues/257

This commit makes the minimal amount of changes to switch us from
whyrusleeping/mdns to grandcat/zeroconf. Of note, rather than asking the
`host.Host` which `addrs` we can listen on, we push this to the
zeroconf library (which runs across available ifaces).

I haven't had a chance to test this on something complicated to see if it retains the desired behavior (to be honest, still new to this codebase, unsure of what all of those would be!), but tests pass locally. 

TODO: 
 - [x] Revert PORT behavior
 - [x] ensure that grandcat/zerconf dep is in gx; tests pass

rargulati commented on p2p/discovery/mdns.go:
```diff
@@ -152,8 +120,13 @@ func (m *mdnsService) pollForEntries(ctx context.Context) {
+	info := strings.Join(e.Text, "|")
+
+	// addripv4 potential issue
+	log.Debugf("Handling MDNS entry: %s:%d %s", e.AddrIPv4[0], e.Port, info)
```
This is the only decision I wasn't certain of.

Kubuxu commented on p2p/discovery/mdns.go:
```diff
@@ -111,34 +75,38 @@ func NewMdnsService(ctx context.Context, peerhost host.Host, interval time.Durat
 
 	ticker := time.NewTicker(m.interval)
 	for {
+		resolver, err := mdns.NewResolver(nil)
```
I think it might be beast to put resolver in `mdnsService` struct and reuse it, instead of creating a new one every iteration.

rargulati replied:
Yikes, good catch. Fixed.

Kubuxu commented on p2p/discovery/mdns.go:
```diff
@@ -34,72 +34,36 @@ type Notifee interface {
 	golog.SetOutput(ioutil.Discard)
 
-	var ipaddrs []net.IP
 	port := 4001
```
This is old but it is very wrong.

Kubuxu commented on p2p/discovery/mdns.go:
```diff
@@ -165,7 +138,7 @@ func (m *mdnsService) handleEntry(e *mdns.ServiceEntry) {
 
 	maddr, err := manet.FromNetAddr(&net.TCPAddr{
-		IP:   e.AddrV4,
+		IP:   e.AddrIPv4[0],
```
As there is possibility of multiple addresses, I would iterate over them, crate maddr for each and add it to the peerstore.

rargulati commented:
@Stebalien @Kubuxu I've addressed the feedback in c92f1f6

In regards to the port, what would you recommend here? Skimmed through https://tools.ietf.org/html/rfc6762 and see that 5353 is used for the listener, but not seeing anything for the port that's announcing.

I notice that I'm failing tests due to a gx issue. Any recommendations? I'll need to learn that right quick :)

Kubuxu commented:
@rargulati It was my mistake, I thought it was the port we were announcing to the world but it is mDNS discovery port.

@Stebalien correct if I am wrong, but currently mDNS only publishes peerID in local network, not the full multiaddr, is that right? If it is so, is there a reason not to publish maddr so we can connect to it? Publishing just peerID sees little fruitless as it won't work without DHT capabilities (in case that bootstrap didn't work).

Stebalien commented:
Port *is* the port of the service. Also, we do announce the IP addresses and we convert port + IP into `/ip4/IP/tcp/PORT`. However, that's... suboptimal. It doesn't account for any other potential transports.

Note: For *now*, I'd be fine preserving the existing behavior. The primary concern here is upgrading to a *working* mDNS library (one that doesn't have memory safety issues).

However, we *also* have a bunch of bugs/inconsistencies that we need to fix. I've written them up here:https://github.com/libp2p/libp2p/issues/28

rargulati commented:
Just to clarify on final things:
-[ ] Revert PORT to 4001 - the fix to this and other issues will be addressed as detailed in libp2p/libp2p#28
-[ ] Get grandcat/zerconf into gx (I'm struggling with this - how do I get a new hash for a new dependency - either of you mind doing a quick how-to in irc/email/here?) 

Does that sound about right @Stebalien @Kubuxu - I'll get this across the line asap.

Stebalien commented:
>  Revert PORT to 4001 - the fix to this and other issues will be addressed as detailed in libp2p/libp2p#28

Actually, restore the previous "figure out what port we're using" code. It's not perfect, but it's better than just assuming port 4001.

>  Get grandcat/zerconf into gx (I'm struggling with this - how do I get a new hash for a new dependency - either of you mind doing a quick how-to in irc/email/here?)

I'll import this into gx tomorrow. We need to write some proper tutorials but it's on our never ending backlog.

Stebalien commented:
> I'll import this into gx tomorrow. We need to write some proper tutorials but it's on our never ending backlog.

This will actually require updating some of our dependencies so it may take a week (blocked on a bunch of other stuff).

deckbsd commented:
Hi, do you know when this commit will be done ?

Stebalien commented:
It's been blocked on:

* [x] Cutting a new release (0.4.14)
* [ ] Fixing a namesys issue: https://github.com/ipfs/go-ipfs/pull/4506
* [ ] Updating the datastores and the DHT (in go-ipfs).

Once those are done, I'll finally be able to bubble a few low-level updates through the entire tree. The problem here is that this zeroconf library requires a newer version of `golang.org/x/net` than we use in go-ipfs.

Stebalien commented:
Still need to fix https://github.com/libp2p/go-libp2p/issues/376 but I'd like to punt on that.

Stebalien commented:
Waiting on https://github.com/cenkalti/backoff/pull/59

Stebalien commented on p2p/discovery/mdns.go:
```diff
@@ -44,62 +45,76 @@ type mdnsService struct {
-	var out []*net.TCPAddr
-	for _, addr := range ph.Addrs() {
+// Tries to pick the best port.
+func getBestPort(addrs []ma.Multiaddr) (int, error) {
```
This should be fixed by https://github.com/libp2p/go-libp2p/issues/376

bigs commented on p2p/discovery/mdns.go:
```diff
@@ -152,8 +165,10 @@ func (m *mdnsService) pollForEntries(ctx context.Context) {
-	log.Debugf("Handling MDNS entry: %s:%d %s", e.AddrV4, e.Port, e.Info)
-	mpeer, err := peer.IDB58Decode(e.Info)
+	// pull out the txt
+	info := strings.Join(e.Text, "|")
```
is this designed to fail if there are less than or more than one entry?

bigs replied:
to be clear, because of the decoding afterwards

Stebalien replied:
Hm. Yeah, that looks wrong. We should just take the first entry and ignore the rest.

Stebalien commented:
Blocked on https://github.com/grandcat/zeroconf/issues/38. Fixing that was the reason we went down this road in the first place so there's no real reason to merge this till we get that fixed.

Stebalien commented:
Given https://github.com/libp2p/specs/pull/80, we should probably just abort this whole thing and implement our own library that:

1. Implements that spec.
2. Doesn't have random race conditions.
3. Can handle things like network interface changes.
